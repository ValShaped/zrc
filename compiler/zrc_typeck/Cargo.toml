[package]
name = "zrc_typeck"
version = "0.1.0"
edition = "2024"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
derive_more = { version = "2.0.1", features = ["display"] }
indexmap = "2.2.6"
zrc_diagnostics = { version = "0.1.0", path = "../zrc_diagnostics" }
zrc_parser = { version = "0.1.0", path = "../zrc_parser" }
zrc_utils = { version = "0.1.0", path = "../zrc_utils" }
