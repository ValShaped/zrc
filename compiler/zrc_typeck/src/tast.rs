//! Type-checked AST for the Zirco programming language
//!
//! This is similar to the [standard AST](zrc_parser::ast), but with type
//! information attached to it.

pub mod expr;
pub mod stmt;
pub mod ty;
