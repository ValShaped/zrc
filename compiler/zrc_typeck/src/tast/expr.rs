//! Expression representation for the Zirco [TAST](super)

use std::fmt::Display;

/// An [expression kind](TypedExprKind) with its yielded [result type](super::ty::Type) attached to it.
#[derive(PartialEq, Debug, Clone)]
#[allow(clippy::module_name_repetitions)]
pub struct TypedExpr(pub super::ty::Type, pub TypedExprKind);

/// The kind of a [`TypedExpr`]
#[derive(PartialEq, Debug, Clone)]
pub enum TypedExprKind {
    /// `a, b`
    Comma(Box<TypedExpr>, Box<TypedExpr>),

    /// `a = b`
    Assignment(Box<TypedExpr>, Box<TypedExpr>),

    /// `!x`
    UnaryNot(Box<TypedExpr>),
    /// `~x`
    UnaryBitwiseNot(Box<TypedExpr>),
    /// `-x`
    UnaryMinus(Box<TypedExpr>),
    /// `&x`
    UnaryAddressOf(Box<TypedExpr>),
    /// `*x`
    UnaryDereference(Box<TypedExpr>),

    /// `a[b]`
    Index(Box<TypedExpr>, Box<TypedExpr>),
    /// `a.b`
    Dot(Box<TypedExpr>, String),
    /// `a->b`
    Arrow(Box<TypedExpr>, String),
    /// `a(b, c, d, ...)`
    Call(Box<TypedExpr>, Vec<TypedExpr>),

    /// `a ? b : c`
    Ternary(Box<TypedExpr>, Box<TypedExpr>, Box<TypedExpr>),

    /// `a && b`
    LogicalAnd(Box<TypedExpr>, Box<TypedExpr>),
    /// `a || b`
    LogicalOr(Box<TypedExpr>, Box<TypedExpr>),

    /// `a == b`
    Equals(Box<TypedExpr>, Box<TypedExpr>),
    /// `a != b`
    NotEquals(Box<TypedExpr>, Box<TypedExpr>),

    /// `a & b`
    BitwiseAnd(Box<TypedExpr>, Box<TypedExpr>),
    /// `a | b`
    BitwiseOr(Box<TypedExpr>, Box<TypedExpr>),
    /// `a ^ b`
    BitwiseXor(Box<TypedExpr>, Box<TypedExpr>),

    /// `a > b`
    GreaterThan(Box<TypedExpr>, Box<TypedExpr>),
    /// `a >= b`
    GreaterThanOrEqualTo(Box<TypedExpr>, Box<TypedExpr>),
    /// `a < b`
    LessThan(Box<TypedExpr>, Box<TypedExpr>),
    /// `a <= b`
    LessThanOrEqualTo(Box<TypedExpr>, Box<TypedExpr>),

    /// `a >> b`
    BitwiseRightShift(Box<TypedExpr>, Box<TypedExpr>),
    /// `a << b`
    BitwiseLeftShift(Box<TypedExpr>, Box<TypedExpr>),

    /// `a + b`
    Addition(Box<TypedExpr>, Box<TypedExpr>),
    /// `a - b`
    Subtraction(Box<TypedExpr>, Box<TypedExpr>),

    /// `a * b`
    Multiplication(Box<TypedExpr>, Box<TypedExpr>),
    /// `a / b`
    Division(Box<TypedExpr>, Box<TypedExpr>),
    /// `a % b`
    Modulo(Box<TypedExpr>, Box<TypedExpr>),

    /// `x as T`
    Cast(Box<TypedExpr>, super::ty::Type),

    /// Any numeric literal.
    NumberLiteral(String),
    /// Any string literal.
    StringLiteral(String),
    /// Any identifier.
    Identifier(String),
    /// Any boolean literal.
    BooleanLiteral(bool),
}

impl Display for TypedExprKind {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::NumberLiteral(n) => write!(f, "{n}"),
            Self::StringLiteral(s) => write!(f, "{s}"),
            Self::Identifier(i) => write!(f, "{i}"),
            Self::BooleanLiteral(b) => write!(f, "{b}"),
            Self::Assignment(l, r) => write!(f, "{l} = {r}"),
            Self::Addition(l, r) => write!(f, "{l} + {r}"),
            Self::Subtraction(l, r) => write!(f, "{l} - {r}"),
            Self::Multiplication(l, r) => write!(f, "{l} * {r}"),
            Self::Division(l, r) => write!(f, "{l} / {r}"),
            Self::Modulo(l, r) => write!(f, "{l} % {r}"),
            Self::BitwiseAnd(l, r) => write!(f, "{l} & {r}"),
            Self::BitwiseOr(l, r) => write!(f, "{l} | {r}"),
            Self::BitwiseXor(l, r) => write!(f, "{l} ^ {r}"),
            Self::BitwiseLeftShift(l, r) => write!(f, "{l} << {r}"),
            Self::BitwiseRightShift(l, r) => write!(f, "{l} >> {r}"),
            Self::GreaterThan(l, r) => write!(f, "{l} > {r}"),
            Self::GreaterThanOrEqualTo(l, r) => write!(f, "{l} >= {r}"),
            Self::LessThan(l, r) => write!(f, "{l} < {r}"),
            Self::LessThanOrEqualTo(l, r) => write!(f, "{l} <= {r}"),
            Self::Equals(l, r) => write!(f, "{l} == {r}"),
            Self::NotEquals(l, r) => write!(f, "{l} != {r}"),
            Self::LogicalAnd(l, r) => write!(f, "{l} && {r}"),
            Self::LogicalOr(l, r) => write!(f, "{l} || {r}"),
            Self::Comma(l, r) => write!(f, "{l}, {r}"),
            Self::UnaryNot(e) => write!(f, "!{e}"),
            Self::Cast(x, t) => write!(f, "{x} as {t}"),
            Self::UnaryBitwiseNot(e) => write!(f, "~{e}"),
            Self::UnaryMinus(e) => write!(f, "-{e}"),
            Self::UnaryAddressOf(e) => write!(f, "&{e}"),
            Self::UnaryDereference(e) => write!(f, "*{e}"),
            Self::Arrow(l, r) => write!(f, "{l}->{r}"),
            Self::Ternary(a, b, c) => write!(f, "{a} ? {b} : {c}"),
            Self::Index(a, b) => write!(f, "{a}[{b}]"),
            Self::Dot(a, b) => write!(f, "{a}.{b}"),
            Self::Call(a, b) => write!(
                f,
                "{a}({})",
                b.iter()
                    .map(ToString::to_string)
                    .collect::<Vec<String>>()
                    .join(", ")
            ),
        }
    }
}

impl Display for TypedExpr {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "(({}) as ({}))", self.1, self.0)
    }
}
