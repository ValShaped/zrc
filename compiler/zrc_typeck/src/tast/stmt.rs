//! Statement representation for the Zirco [TAST](super)

use std::fmt::Display;

use indexmap::IndexMap;

use super::{expr::TypedExpr, ty::Type};

/// A declaration created with `let`.
#[derive(Debug, Clone, PartialEq)]
pub struct LetDeclaration {
    /// The name of the identifier.
    pub name: String,
    /// The type of the new symbol. If set to [`None`], the type will be
    /// inferred.
    pub ty: Type, // types are definite after inference
    /// The value to associate with the new symbol.
    pub value: Option<TypedExpr>,
}

impl Display for LetDeclaration {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match &self.value {
            Some(v) => write!(f, "{}: {} = {}", self.name, self.ty, v),
            None => write!(f, "{}: {}", self.name, self.ty),
        }
    }
}

/// The enum representing all of the different kinds of statements in Zirco
/// after type checking
#[derive(Debug, Clone, PartialEq)]
#[allow(clippy::module_name_repetitions)]
pub enum TypedStmt {
    // all of the Box<Stmt>s for "possibly blocks" have been desugared into vec[single stmt] here
    // (basically if (x) y has become if (x) {y})
    /// `if (x) y` or `if (x) y else z`
    IfStmt(TypedExpr, Vec<TypedStmt>, Option<Vec<TypedStmt>>),
    /// `while (x) y`
    WhileStmt(TypedExpr, Vec<TypedStmt>),
    /// `for (init; cond; post) body`
    ForStmt {
        /// Runs once before the loop starts.
        init: Option<Box<Vec<LetDeclaration>>>,
        /// Runs before each iteration of the loop. If this evaluates to
        /// `false`, the loop will end. If this is [`None`], the loop
        /// will run forever.
        cond: Option<TypedExpr>,
        /// Runs after each iteration of the loop.
        post: Option<TypedExpr>,
        /// The body of the loop.
        body: Vec<TypedStmt>,
    },
    /// `{ ... }`
    BlockStmt(Vec<TypedStmt>),
    /// `x;`
    ExprStmt(TypedExpr),
    /// `;`
    EmptyStmt,
    /// `continue;`
    ContinueStmt,
    /// `break;`
    BreakStmt,
    /// `return;` or `return x;`
    ReturnStmt(Option<TypedExpr>),
    /// A let declaration
    DeclarationList(Vec<LetDeclaration>),
}

/// A struct or function declaration at the top level of a file
#[derive(Debug, Clone, PartialEq)]
pub enum TypedDeclaration {
    /// A declaration of a function
    FunctionDeclaration {
        /// The name of the function.
        name: String,
        /// The parameters of the function.
        parameters: Vec<ArgumentDeclaration>,
        /// The return type of the function. If set to [`None`], the function is
        /// void.
        return_type: Option<Type>,
        /// The body of the function. If set to [`None`], this is an extern
        /// declaration.
        body: Option<Vec<TypedStmt>>,
    },
    /// A named declaration for a `struct`.
    StructDeclaration {
        /// The name of the newtype.
        name: String,
        /// The key-value pairs of the struct. Ordered by declaration order.
        fields: IndexMap<String, super::ty::Type>,
    },
}

impl Display for TypedDeclaration {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::FunctionDeclaration {
                name,
                parameters,
                return_type: Some(r),
                body: Some(body),
            } => write!(
                f,
                "fn {name}({}) -> {r} {{\n{}\n}}",
                parameters
                    .iter()
                    .map(ToString::to_string)
                    .collect::<Vec<String>>()
                    .join(", "),
                body.iter()
                    .map(|stmt| stmt
                        .to_string()
                        .split('\n')
                        .map(|x| format!("    {x}"))
                        .collect::<Vec<_>>()
                        .join("\n"))
                    .collect::<Vec<String>>()
                    .join("\n")
            ),
            Self::FunctionDeclaration {
                name,
                parameters,
                return_type: Some(r),
                body: None,
            } => write!(
                f,
                "fn {name}({}) -> {r};",
                parameters
                    .iter()
                    .map(ToString::to_string)
                    .collect::<Vec<String>>()
                    .join(", ")
            ),
            Self::FunctionDeclaration {
                name,
                parameters,
                return_type: None,
                body: Some(body),
            } => write!(
                f,
                "fn {name}({}) {{\n{}\n}}",
                parameters
                    .iter()
                    .map(ToString::to_string)
                    .collect::<Vec<String>>()
                    .join(", "),
                body.iter()
                    .map(|stmt| stmt
                        .to_string()
                        .split('\n')
                        .map(|x| format!("    {x}"))
                        .collect::<Vec<_>>()
                        .join("\n"))
                    .collect::<Vec<String>>()
                    .join("\n")
            ),
            Self::FunctionDeclaration {
                name,
                parameters,
                return_type: None,
                body: None,
            } => write!(
                f,
                "fn {name}({});",
                parameters
                    .iter()
                    .map(ToString::to_string)
                    .collect::<Vec<String>>()
                    .join(", ")
            ),
            Self::StructDeclaration { name, fields } => write!(
                f,
                "struct {name} {{\n{}\n}}",
                fields
                    .iter()
                    .map(|(name, ty)| format!("    {name}: {ty}"))
                    .collect::<Vec<_>>()
                    .join(",\n")
            ),
        }
    }
}

impl Display for TypedStmt {
    #[allow(clippy::too_many_lines)]
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::IfStmt(e, s1, Some(s2)) => write!(
                f,
                "if ({e}) {{\n{}\n}} else {{\n{}\n}}",
                s1.iter()
                    .map(|stmt| stmt
                        .to_string()
                        .split('\n')
                        .map(|x| format!("    {x}"))
                        .collect::<Vec<_>>()
                        .join("\n"))
                    .collect::<Vec<_>>()
                    .join("\n"),
                s2.iter()
                    .map(|stmt| stmt
                        .to_string()
                        .split('\n')
                        .map(|x| format!("    {x}"))
                        .collect::<Vec<_>>()
                        .join("\n"))
                    .collect::<Vec<_>>()
                    .join("\n")
            ),
            Self::IfStmt(e, s1, None) => write!(
                f,
                "if ({e}) {{\n{}\n}}",
                s1.iter()
                    .map(|stmt| stmt
                        .to_string()
                        .split('\n')
                        .map(|x| format!("    {x}"))
                        .collect::<Vec<_>>()
                        .join("\n"))
                    .collect::<Vec<_>>()
                    .join("\n")
            ),
            Self::WhileStmt(e, s) => write!(
                f,
                "while ({e}) {{\n{}\n}}",
                s.iter()
                    .map(|stmt| stmt
                        .to_string()
                        .split('\n')
                        .map(|x| format!("    {x}"))
                        .collect::<Vec<_>>()
                        .join("\n"))
                    .collect::<Vec<_>>()
                    .join("\n")
            ),
            Self::ForStmt {
                init,
                cond,
                post,
                body,
            } => write!(
                f,
                "for ({} {}; {}) {{\n{}\n}}",
                init.clone().map_or(String::new(), |x| format!(
                    "let {};",
                    x.iter()
                        .map(std::string::ToString::to_string)
                        .collect::<Vec<_>>()
                        .join(", ")
                )),
                cond.clone().map_or(String::new(), |x| x.to_string()),
                post.clone().map_or(String::new(), |x| x.to_string()),
                body.iter()
                    .map(|stmt| stmt
                        .to_string()
                        .split('\n')
                        .map(|x| format!("    {x}"))
                        .collect::<Vec<_>>()
                        .join("\n"))
                    .collect::<Vec<_>>()
                    .join("\n")
            ),
            Self::BlockStmt(s) => write!(
                f,
                "{{\n{}\n}}",
                s.iter()
                    .map(|stmt| stmt
                        .to_string()
                        .split('\n')
                        .map(|x| format!("    {x}"))
                        .collect::<Vec<_>>()
                        .join("\n"))
                    .collect::<Vec<_>>()
                    .join("\n")
            ),
            Self::ExprStmt(e) => write!(f, "{e};"),
            Self::EmptyStmt => write!(f, ";"),
            Self::ContinueStmt => write!(f, "continue;"),
            Self::BreakStmt => write!(f, "break;"),
            Self::ReturnStmt(Some(e)) => write!(f, "return {e};",),
            Self::ReturnStmt(None) => write!(f, "return;"),
            Self::DeclarationList(d) => write!(
                f,
                "let {};",
                d.iter()
                    .map(std::string::ToString::to_string)
                    .collect::<Vec<_>>()
                    .join(", ")
            ),
        }
    }
}

/// A special form of [`LetDeclaration`] used for function parameters.
#[derive(PartialEq, Debug, Clone)]
pub struct ArgumentDeclaration {
    /// The name of the parameter.
    pub name: String,
    /// The type of the parameter.
    pub ty: Type,
}
impl Display for ArgumentDeclaration {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}: {}", self.name, self.ty)
    }
}
