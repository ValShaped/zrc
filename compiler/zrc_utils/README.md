# common utilities used in `zrc`

This crate exists to bring some of the utility types like [`span::Spanned`] and [`span::Span`] to
a common place to prevent circular dependencies on things like `zrc_parser` (where the `Spanned`
type used to be declared).

Read the module-level documentation for more details.
