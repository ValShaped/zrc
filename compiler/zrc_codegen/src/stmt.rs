//! Code generation for statements
//!
//! This module contains functions to generate LLVM IR for
//! [`zrc_typeck::tast::stmt::TypedStmt`] statements, which
//! represent all statements in the Zirco language.
//!
//! The main function is [`cg_block`], which takes a vector of
//! `TypedStmt`s (a block) and generates the corresponding LLVM IR
//! to execute the statements in order.
//!
//! It also contains helper functions for generating specific
//! statement types, such as [`cg_let_declaration`]
//! which handles `let` declarations.
//!
//! The code generator maintains a [`CgScope`] to track variable
//! bindings and their corresponding LLVM values.

mod branch;
mod let_decl;
mod loops;
mod switch;

use inkwell::{
    basic_block::BasicBlock,
    debug_info::{AsDIScope, DILexicalBlock},
};
use zrc_typeck::tast::{
    stmt::{TypedStmt, TypedStmtKind},
    ty::Type,
};
use zrc_utils::span::{Spannable, Spanned};

use crate::{
    ctx::{BlockCtx, FunctionCtx},
    expr::cg_expr,
    scope::CgScope,
    ty::llvm_basic_type,
};

/// Consists of the [`BasicBlock`]s to `br` to when encountering certain
/// instructions. It is passed to [`cg_block`] to allow it to properly handle
/// break and continue.
#[derive(PartialEq, Eq, Debug, Clone)]
#[allow(clippy::redundant_pub_crate)]
pub(crate) struct LoopBreakaway<'ctx> {
    /// Points to the exit basic block.
    on_break: BasicBlock<'ctx>,
    /// For `for` loops, points to the latch. For `while` loops, points to the
    /// header.
    on_continue: BasicBlock<'ctx>,
}

/// Process a vector of [`TypedStmt`]s (a block) and handle each statement.
///
/// # Panics
/// Panics if an internal code generation error is encountered.
#[allow(
    clippy::too_many_arguments,
    clippy::too_many_lines,
    clippy::needless_pass_by_value,
    clippy::redundant_pub_crate,
    clippy::ref_option
)]
pub(crate) fn cg_block<'ctx, 'input, 'a>(
    cg: FunctionCtx<'ctx, 'a>,
    bb: BasicBlock<'ctx>,
    parent_scope: &'a CgScope<'input, 'ctx>,
    parent_lexical_block: DILexicalBlock<'ctx>,
    block: Spanned<Vec<TypedStmt<'input>>>,
    breakaway: &Option<LoopBreakaway<'ctx>>,
) -> Option<BasicBlock<'ctx>> {
    let mut scope = parent_scope.clone();
    let block_span = block.span();
    let block_line_col = cg.line_lookup.lookup_from_index(block_span.start());
    let lexical_block = cg.dbg_builder.create_lexical_block(
        parent_lexical_block.as_debug_info_scope(),
        cg.compilation_unit.get_file(),
        block_line_col.line,
        block_line_col.col,
    );

    block
        .into_value()
        .into_iter()
        .try_fold(bb, |bb, stmt| -> Option<BasicBlock> {
            let stmt_span = stmt.0.span();
            let stmt_line_col = cg.line_lookup.lookup_from_index(stmt_span.start());
            let debug_location = cg.dbg_builder.create_debug_location(
                cg.ctx,
                stmt_line_col.line,
                stmt_line_col.col,
                lexical_block.as_debug_info_scope(),
                None,
            );
            cg.builder.set_current_debug_location(debug_location);

            match stmt.0.into_value() {
                TypedStmtKind::UnreachableStmt => {
                    cg.builder
                        .build_unreachable()
                        .expect("unreachable should generate successfully");

                    None
                }

                TypedStmtKind::SwitchCase {
                    scrutinee,
                    default,
                    cases,
                } => Some(switch::cg_switch_stmt(
                    cg,
                    bb,
                    &scope,
                    lexical_block,
                    breakaway,
                    stmt_span,
                    scrutinee,
                    default,
                    cases,
                )),

                TypedStmtKind::ExprStmt(expr) => {
                    let expr_cg = BlockCtx::new(cg, &scope, lexical_block);

                    Some(cg_expr(expr_cg, bb, expr).bb)
                }

                TypedStmtKind::IfStmt(cond, then, then_else) => branch::cg_if_stmt(
                    cg,
                    bb,
                    &scope,
                    lexical_block,
                    breakaway,
                    cond,
                    then,
                    then_else,
                ),

                TypedStmtKind::BlockStmt(block) => cg_block(
                    cg,
                    bb,
                    &scope,
                    lexical_block,
                    block.in_span(stmt_span),
                    breakaway,
                ),

                TypedStmtKind::ReturnStmt(Some(expr)) => {
                    let expr_cg = BlockCtx::new(cg, &scope, lexical_block);

                    let expr = cg_expr(expr_cg, bb, expr).into_value();

                    cg.builder
                        .build_return(Some(&expr))
                        .expect("return should generate successfully");

                    None
                }

                TypedStmtKind::ReturnStmt(None) => {
                    let unit_type = llvm_basic_type(&cg, &Type::unit());

                    cg.builder
                        .build_return(Some(&unit_type.0.const_zero()))
                        .expect("return should generate successfully");

                    None
                }

                TypedStmtKind::ContinueStmt => {
                    cg.builder
                        .build_unconditional_branch(
                            breakaway
                                .as_ref()
                                .expect("`breakaway` should exist all places `continue` is valid")
                                .on_continue,
                        )
                        .expect("branch should generate successfully");

                    None
                }

                TypedStmtKind::BreakStmt => {
                    cg.builder
                        .build_unconditional_branch(
                            breakaway
                                .as_ref()
                                .expect("`breakaway` should exist all places `break` is valid")
                                .on_break,
                        )
                        .expect("branch should generate successfully");

                    None
                }

                TypedStmtKind::DeclarationList(declarations) => Some(let_decl::cg_let_declaration(
                    cg,
                    bb,
                    &mut scope,
                    lexical_block,
                    declarations,
                )),

                TypedStmtKind::ForStmt {
                    init,
                    cond,
                    post,
                    body,
                } => Some(loops::cg_for_stmt(
                    cg,
                    bb,
                    &scope,
                    lexical_block,
                    init,
                    cond,
                    post,
                    body,
                )),

                TypedStmtKind::WhileStmt(cond, body) => {
                    Some(loops::cg_while_stmt(cg, &scope, lexical_block, cond, body))
                }

                TypedStmtKind::DoWhileStmt(body, cond) => Some(loops::cg_do_while_stmt(
                    cg,
                    &scope,
                    lexical_block,
                    body,
                    cond,
                )),
            }
        })
}

#[cfg(test)]
mod tests {
    // Please read the "Common patterns in tests" section of crate::test_utils for
    // more information on how code generator tests are structured.

    use indoc::indoc;

    use crate::cg_snapshot_test;

    #[test]
    fn unreachable_statement_generates_properly() {
        cg_snapshot_test!(indoc! {"
            fn test(cond: bool) {
                let x = 7;
                if (x == 6) unreachable;
                else return;
            }
        "});
    }
}
