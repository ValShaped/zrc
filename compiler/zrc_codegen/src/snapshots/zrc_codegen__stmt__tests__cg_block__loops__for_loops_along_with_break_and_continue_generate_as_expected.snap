---
source: compiler/zrc_codegen/src/stmt.rs
description: "fn get_int() -> i32;\n\nfn test() {\n    // TEST: the proper while loop structure is created\n    for (let i = 0; i < get_int(); i += 1) {\n        // TEST: break jumps to the `end` block\n        if (i > get_int()) break;\n        else {\n            // TEST: continue jumps to the latch block\n            if (i < get_int()) continue;\n            else {}\n        }\n\n        // TEST: the loop jumps to the latch block which jumps back to the\n        // header\n    }\n}\n"
expression: resulting_ir
---
; ModuleID = 'test.zr'
source_filename = "test.zr"

declare i32 @get_int()

define {} @test() !dbg !3 {
entry:
  %let_i = alloca i32, align 4
  call void @llvm.dbg.declare(metadata ptr %let_i, metadata !9, metadata !DIExpression()), !dbg !13
  store i32 0, ptr %let_i, align 4, !dbg !13
  br label %header, !dbg !13

header:                                           ; preds = %latch, %entry
  %load = load i32, ptr %let_i, align 4, !dbg !14
  %call = call i32 @get_int(), !dbg !15
  %cmp = icmp ult i32 %load, %call, !dbg !15
  br i1 %cmp, label %body, label %exit, !dbg !15

body:                                             ; preds = %header
  %load1 = load i32, ptr %let_i, align 4, !dbg !16
  %call2 = call i32 @get_int(), !dbg !18
  %cmp3 = icmp ugt i32 %load1, %call2, !dbg !18
  br i1 %cmp3, label %then, label %then_else, !dbg !18

latch:                                            ; preds = %end9, %then7
  %load10 = load i32, ptr %let_i, align 4, !dbg !19
  %add = add i32 %load10, 1, !dbg !20
  store i32 %add, ptr %let_i, align 4, !dbg !19
  br label %header, !dbg !19

exit:                                             ; preds = %then, %header
  ret {} zeroinitializer, !dbg !21

then:                                             ; preds = %body
  br label %exit, !dbg !22

then_else:                                        ; preds = %body
  %load4 = load i32, ptr %let_i, align 4, !dbg !24
  %call5 = call i32 @get_int(), !dbg !26
  %cmp6 = icmp ult i32 %load4, %call5, !dbg !26
  br i1 %cmp6, label %then7, label %then_else8, !dbg !26

then7:                                            ; preds = %then_else
  br label %latch, !dbg !27

then_else8:                                       ; preds = %then_else
  br label %end, !dbg !29

end:                                              ; preds = %then_else8
  br label %end9, !dbg !30

end9:                                             ; preds = %end
  br label %latch, !dbg !30
}

; Function Attrs: nocallback nofree nosync nounwind speculatable willreturn memory(none)
declare void @llvm.dbg.declare(metadata, metadata, metadata) #0

attributes #0 = { nocallback nofree nosync nounwind speculatable willreturn memory(none) }

!llvm.module.flags = !{!0}
!llvm.dbg.cu = !{!1}

!0 = !{i32 2, !"Debug Info Version", i32 3}
!1 = distinct !DICompileUnit(language: DW_LANG_C, file: !2, producer: "zrc test runner", isOptimized: false, flags: "zrc --fake-args", runtimeVersion: 0, emissionKind: FullDebug, splitDebugInlining: false)
!2 = !DIFile(filename: "test.zr", directory: "/fake/path")
!3 = distinct !DISubprogram(name: "test", linkageName: "test", scope: null, file: !2, line: 3, type: !4, scopeLine: 3, spFlags: DISPFlagDefinition, unit: !1, retainedNodes: !8)
!4 = !DISubroutineType(types: !5)
!5 = !{!6}
!6 = !DICompositeType(tag: DW_TAG_structure_type, name: "(struct {})", scope: !2, file: !2, elements: !7)
!7 = !{}
!8 = !{!9}
!9 = !DILocalVariable(name: "i", scope: !10, file: !2, line: 5, type: !12)
!10 = distinct !DILexicalBlock(scope: !11, file: !2, line: 3, column: 11)
!11 = distinct !DILexicalBlock(scope: !3, file: !2, line: 3, column: 11)
!12 = !DIBasicType(name: "i32")
!13 = !DILocation(line: 5, column: 14, scope: !10)
!14 = !DILocation(line: 5, column: 21, scope: !10)
!15 = !DILocation(line: 5, column: 25, scope: !10)
!16 = !DILocation(line: 7, column: 13, scope: !17)
!17 = distinct !DILexicalBlock(scope: !10, file: !2, line: 5, column: 44)
!18 = !DILocation(line: 7, column: 17, scope: !17)
!19 = !DILocation(line: 5, column: 36, scope: !10)
!20 = !DILocation(line: 5, column: 41, scope: !10)
!21 = !DILocation(line: 17, column: 1, scope: !10)
!22 = !DILocation(line: 7, column: 28, scope: !23)
!23 = distinct !DILexicalBlock(scope: !17, file: !2, line: 7, column: 28)
!24 = !DILocation(line: 10, column: 17, scope: !25)
!25 = distinct !DILexicalBlock(scope: !17, file: !2, line: 8, column: 14)
!26 = !DILocation(line: 10, column: 21, scope: !25)
!27 = !DILocation(line: 10, column: 32, scope: !28)
!28 = distinct !DILexicalBlock(scope: !25, file: !2, line: 10, column: 32)
!29 = !DILocation(line: 10, column: 41, scope: !25)
!30 = !DILocation(line: 7, column: 34, scope: !17)
